python/lsst/ip/isr/isrMockLSST.py

```diff
@@ -129,139 +223,227 @@ def makeImage(self):
-                self.amplifierAddFringe(amp, ampData, np.array(self.config.fringeScale) * self.config.gain,
+                # Fringes are added in electrons.
+                self.amplifierAddFringe(amp,
+                                        ampImageData,
+                                        np.array(self.config.fringeScale),
                                         x0=np.array(self.config.fringeX0),
                                         y0=np.array(self.config.fringeY0))
 
             if self.config.doAddFlat:
                 if self.config.calibMode:
                     # In case we are making a combined flat,
                     # add a non-zero signal so the mock flat can be multiplied
-                    self.amplifierAddNoise(ampData, 1.0, 0.0)
+                    self.amplifierAddNoise(ampImageData, 1.0, 0.0)
                 # Multiply each amplifier by a Gaussian centered on u0 and v0
                 u0 = exposure.getDetector().getBBox().getDimensions().getX()/2.
                 v0 = exposure.getDetector().getBBox().getDimensions().getY()/2.
-                self.amplifierMultiplyFlat(amp, ampData, self.config.flatDrop, u0=u0, v0=v0)
-
-            # ISR effects
-            # 1. Add dark in e- (darkRate is configured in e-/s)
-            # TODO: DM-42880 to set configs to correct units
-            if self.config.doAddDark:
-                self.amplifierAddNoise(ampData,
-                                       self.config.darkRate * self.config.darkTime,
-                                       0. if self.config.calibMode
-                                       else np.sqrt(self.config.darkRate * self.config.darkTime))
-
-            # 2. Gain normalize  (from e- to ADU)
-            # TODO: DM-43601 gain from PTC per amplifier
-            # TODO: DM-36639 gain with temperature dependence
-            if self.config.doApplyGain:
-                self.applyGain(ampData, self.config.gain)
+                self.amplifierMultiplyFlat(amp, ampImageData, self.config.flatDrop, u0=u0, v0=v0)
+
+            # On-chip electronic effects.
 
-            # 3. Add read noise to the image region in ADU.
+            # TODO
+            # 1. Add BF effect (e-)
+
+            # 2. Add dark current (e-) to imaging portion of the amplifier.
+            if self.config.doAddDark or self.config.doAddDarkNoiseOnly:
+                if self.config.doAddDarkNoiseOnly:
+                    darkLevel = 0.0
+                else:
+                    darkLevel = self.config.darkRate * self.config.darkTime
+                if self.config.calibMode:
+                    darkNoise = 0.0
+                else:
+                    darkNoise = np.sqrt(self.config.darkRate * self.config.darkTime)
+
+                self.amplifierAddNoise(ampImageData, darkLevel, darkNoise, rng=rngDark)
+
+            # 3. Add 2D bias residual (e-) to imaging portion of the amplifier.
+            if self.config.doAdd2DBias:
```

Comment:
Yes that's true, it's an unstructured noise field so that we can have something that is a "2D bias residual pattern" even if it isn't an accurate one, so that we can test that it get removes (and not just subtracting zero).